File: ios/Analytics/Scrubber.swift
import Foundation
import Sentry

enum Scrubber {
    private static let maxFrames = 20

    static func scrub(event: Event?) -> Event? {
        guard let event else { return nil }
        event.user = nil
        event.request = nil
        event.serverName = nil
        if var contexts = event.context, !contexts.isEmpty {
            contexts.removeValue(forKey: "device")
            contexts.removeValue(forKey: "trace")
            event.context = contexts
        }
        if var breadcrumbs = event.breadcrumbs, !breadcrumbs.isEmpty {
            breadcrumbs = breadcrumbs.filter { breadcrumb in
                guard let message = breadcrumb.message?.lowercased() else { return true }
                return !(message.contains("@") || message.contains("email") || message.contains("ssn"))
            }
            event.breadcrumbs = Array(breadcrumbs.prefix(30))
        }
        if let exceptions = event.exceptions {
            for exception in exceptions {
                if var frames = exception.stacktrace?.frames, frames.count > maxFrames {
                    frames = Array(frames.suffix(maxFrames))
                    exception.stacktrace?.frames = frames
                }
            }
        }
        event.extra = [:]
        return event
    }
}

File: android/app/src/main/java/com/golfiq/hud/analytics/Scrubber.kt
package com.golfiq.hud.analytics

import io.sentry.Breadcrumb
import io.sentry.SentryEvent
import io.sentry.protocol.SentryException
import io.sentry.protocol.SentryStackTrace

object Scrubber {
    private const val MAX_STACK_FRAMES = 20

    fun scrub(event: SentryEvent?): SentryEvent? {
        if (event == null) {
            return null
        }
        event.user = null
        event.request = null
        event.serverName = null
        event.contexts.remove("device")
        event.contexts.remove("app")
        event.contexts.remove("trace")
        event.breadcrumbs = event.breadcrumbs
            ?.filterNot { containsPii(it) }
            ?.take(30)
        event.exceptions?.forEach { exception: SentryException ->
            val stackTrace = exception.stacktrace ?: return@forEach
            trimStackTrace(stackTrace)
        }
        event.extra.clear()
        return event
    }

    private fun containsPii(breadcrumb: Breadcrumb): Boolean {
        val message = buildString {
            append(breadcrumb.message ?: "")
            if (!breadcrumb.data.isNullOrEmpty()) {
                append(' ')
                append(breadcrumb.data.values.joinToString(separator = " "))
            }
        }
        val lowered = message.lowercase()
        return lowered.contains("@") || lowered.contains("email") || lowered.contains("ssn")
    }

    private fun trimStackTrace(stackTrace: SentryStackTrace) {
        val frames = stackTrace.frames ?: return
        if (frames.size > MAX_STACK_FRAMES) {
            stackTrace.frames = frames.takeLast(MAX_STACK_FRAMES)
        }
    }
}
